### PR TITLE
tolist to return multidimensional list

### PR DESCRIPTION
**Pull request by geohotstan**
I tried hacking around indexing the memoryview and reconstructing the list with the right shape for awhile, only to find out that `memoryview.tolist()` exists 😄 

is this method valid?

after casting the memoryview, it seems that tolist resolves to a multidimensional list

### CODE REVIEW

**Comment by chenyuxyz** [comment]
why is the return type not `List`?

**Comment by geohotstan** [comment]
> why is the return type not `List`?

mypy failure:
```
tinygrad/tensor.py:195: error: Incompatible return value type (got "list[int]", expected "list[float | int | bool]")  [return-value]
tinygrad/tensor.py:195: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
tinygrad/tensor.py:195: note: Consider using "Sequence" instead, which is covariant
```

for some reason memoryview's tolist typing is `def tolist(self) -> list[int]: ...`
even when memoryview.tolist can return non int values
image

🤔

**Comment by geohot** [comment]
LGTM, any reason this is draft?

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -191,13 +191,13 @@ def _data(self) -> memoryview:
   def data(self) -> memoryview:
     assert self.dtype.fmt is not None, f"no fmt dtype for {self.dtype}"
     assert all_int(self.shape), f"no data if shape is symbolic, {self.shape=}"
-    return self._data().cast(self.dtype.fmt, self.shape if len(self.shape) else (1,))
+    return self._data().cast(self.dtype.fmt, self.shape)
```
Comment (geohotstan) [inline tinygrad/tensor.py:194]: a little unsure of this change.
But an empty shape `()` works and it matches torch.

Comment (chenyuxyz) [reply tinygrad/tensor.py:194]: do we already have a test for this?

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -191,13 +191,13 @@ def _data(self) -> memoryview:
...
   def item(self) -> ConstType:
     """Returns the value of this tensor as a standard Python number."""
     assert self.dtype.fmt is not None, f"no fmt dtype for {self.dtype}"
     assert self.numel() == 1, "must have one element for item"
     return self._data().cast(self.dtype.fmt)[0]
-  def tolist(self) -> List[ConstType]: return list(self.data())
+  def tolist(self) -> Sequence[ConstType]: return self.data().tolist()
```
Comment (chenyuxyz) [inline tinygrad/tensor.py:200]: add a TODO note pointing to source if it really needs to be `Sequence` for now. this makes mypy happy but user unhappy

File: tinygrad/tensor.py (hunk 1 above)
Comment (geohotstan) [reply tinygrad/tensor.py:194]: There isn't.
I added a test in `test_tensor_data` but I dunno is it's right :s

File: test/test_tensor_data.py [hunk 3]
```diff
@@ -31,6 +31,14 @@ def test_data_nested(self):
     assert dat[0, 0] == 1
     assert dat[1, 1] == 4
 
+  def test_data_const(self):
+    a = Tensor(3, dtype=dtypes.int32)
+    dat = a.data()
+    assert dat.format == "i"
+    assert dat.itemsize == 4
+    assert dat.tolist() == 3
```
Comment (chenyuxyz) [inline test/test_tensor_data.py:39]: slightly weird that `.tolist()` returns a non-list, will let @geohot decide

Comment (geohot) [reply test/test_tensor_data.py:39]: It's fine, it matches PyTorch

**Pull request merged**